### DISCUSSION TITLE
getConfig from 'next/config' doesn't work in _middleware.ts

### DISCUSSION THREAD

**Discussion by lavcraft** [vercel/next.js#35690] [Ideas]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

next 12.1.2
node 16
react 17

### Describe the Bug

Run applicatoin with _middleware like this

```javascript
import getConfig from 'next/config'
import type { NextRequest } from 'next/server'

const config = getConfig();

export async function middleware(req: NextRequest) {
...
}
```

config variable is undefinde, but in pages and getServerSideProps it exists

### Expected Behavior

getConfig return object with serverRuntimeConfig and publicRuntimeConfig

### To Reproduce

1. create next app with typescript
2. add middleware
3. use getConfig from 'next/config' in middleware

**Comment by MauriceArikoglu** [comment]
@lavcraft Middleware Runtime does not allow filesystem access. I believe getConfig is therefore not permitted. See https://nextjs.org/docs/api-reference/edge-runtime for available APIs.

What are you trying to accomplish? Maybe there's alternatives?

**Comment by balazsorban44** [comment]
This is not a bug, only `process.env` is supported at the moment: https://nextjs.org/docs/api-reference/next/server#how-do-i-access-environment-variables

This is to avoid leaking unused environment variables into the runtime.

**Comment by lavcraft** [comment]
I want to reuse my secret and cookieName from serverRuntimeConfig. I guess, this restriction reduce effort from getConfig.

`process.env` - work way. But now, in next.js no any way to standardize configuration management. For example (in framework from another world) - [Externalize Configuration with Spring Boot](https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.external-config)

Main problem with current restrictions:
1. wright next.config.js and parse env and other source manually
2. write it into serverRuntimeConfig
3. obtain in whole application via getConfig
4. write decorated function with right types for use application configuration

But, this restrictions force me to use process.env. From this point, i have two way to configure my application props. That a problem, because no abstraction or way to access to properties. I understand technical restrictions, but not intuitive now and in my opinion, it can be resolved. May be it is a big issue and not easy, but i believe that configuration is important part of any application and simple process.env is not enough

@balazsorban44 okay, instead of close, may be convert it to feature request?

**Comment by lavcraft** [comment]
@balazsorban44 thanks for convert. Hope it will interesting to others

**Comment by gerbyzation** [comment]
Related issue: https://github.com/vercel/next.js/issues/33068

@balazsorban44 I'm a little baffled that these issues are just getting closed and that not having runtime configuration is considered a feature not a bug, but maybe I'm misunderstanding something.

We run nextjs in containers and not having runtime configuration makes nextjs mutually exclusive to building a 12 factor application style containerised application: https://12factor.net/config. As far as I understand pre-building your next application as part of the docker build process isn't an anti-pattern, as the deploy page refers to a docker example which does exactly that: https://github.com/vercel/next.js/blob/canary/examples/with-docker/Dockerfile. Or are you expecting that people ALWAYS build their application for a specific deployment target?

**Comment by laurens-runnable** [comment]
For `getConfig` to work, middleware would have to support the `nodejs` runtime. Currently it only supports the `edge` runtime.

https://github.com/vercel/next.js/discussions/34179
